Linker to NativeAOT sync

vitek-karas opened the pull request:
This brings changes from the linker repo to ilc roughly as of a week ago. This required some non-trivial changes to the code either already in ilc or the one coming from linker:
* Port of CompilerGeneratedState
  * Modify it to use lock free hashtable (since it's a global cache and thus needs to be thread safe)
  * In ILC it requires to operate on generic definitions of everything - so add asserts and code to guarantee this
  * It's not a static class anymore, so pass the instance around as necessary
  * Requires IL access, so needs ILProvider on input
* Cyclic dependency between Logger and CompilerGeneratedState, one needs the instance of the other to work. For now solved by setting it through a property on CompilerGeneratedState
* Split of ReflectionMethodBodyScanner into AttributeDataFlow and GenericParameterDataFlow - mostly direct port from linker
  * The AttributeDataFlow works slightly differently in ilc so adopted the code there
* Changes to DiagnosticContext
  * Can selectively disable Trim, AOT, Single-File diagnostics - used to correctly apply RUC, RDC and RAF suppressions
* Improve handling of RUC and similar warnings in ReflectionMarker
* Added ILOffset to MessageOrigin
  * Even though we already store file/column, the ILOffset is necessary since we use the MessageOrigin as a key in dictionary for recorded patterns and thus we need a way to make the origin unique for each callsite (and file/line/column might not be unique or if there's no PDB those will be null anyway).
* Enable Equality/HashCode implementation on MessageOrigin, still no CompareTo since we don't need it for anything right now.

vitek-karas committed:
First

vitek-karas committed:
Progress

vitek-karas committed:
Small fix

vitek-karas committed:
Improvements to warning suppression and general cleanup of diagnostics handling

vitek-karas committed:
Progress on CompilerGeneratedState

vitek-karas committed:
Finish CompilerGeneratedState port (at least first pass)

vitek-karas committed:
Make the simple stuff work

vitek-karas committed:
Fixes

vitek-karas committed:
Formatting

vitek-karas committed:
Formatting

vitek-karas committed:
Fixes - rewrite compiler generated state

ILC is multi-threaded and CompilerGeneratedState stores a global cache. So the cache has to be synchronized. This rewrites it to use LockFreeReaderHashtable.

vitek-karas committed:
Fixes

With this it passes smoketests

vitek-karas committed:
Cleanup and fixes

vitek-karas commented:
@tlakollo could you please review the changes to CompilerGeneratedState specifically? In general the code should be the same as in linker, but it's now thread safe, so it required quite a few refactorings.

vitek-karas commented:
I made some changes to the linker repo to make this port easier: https://github.com/dotnet/linker/pull/2873

tlakollo reviewed (commented):
Still reviewing but leaving some comments for now

vitek-karas commented:
FYI: I will have to port the most recent changes from Sven over as well - with those it should fix the failures seen in CI. Ultimately we're hitting this bug: https://github.com/dotnet/linker/issues/2874. The linker doesn't crash due to changes from Sven, but NativeAOT currently crashes on it.

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Logging/MessageOrigin.cs:
```diff
@@ -3,45 +3,53 @@
+    public struct MessageOrigin :
 #if false
-        : IComparable<MessageOrigin>, IEquatable<MessageOrigin>
+        IComparable<MessageOrigin>,
```
It was there before, so I kept it for now. I agree that we may eventually remove it, depends on how exact we want to be with the linker. Linker sorts all warnings before printing them out. I think it would make sense to do the same for ILC - even more so, since ILC is multi-threaded and thus more likely to produce warnings in random order. We should catch this as we port more tests over.

MichalStrehovsky replied:
The `#if false` is equivalent to commenting out code. I used `#if false` for larger blocks of code here. I tend to keep commented out code for stuff that consciously wasn't ported over for now.

We track determinism as "the output files should be the same". We haven't really bothered with ordering the warnings. Sometimes it's nice to get warnings from the tool even before the tool finishes. One can just "Ctrl-C" the tool and go fix the warning if it's relevant.

vitek-karas committed:
PR Feedback - just minor tweaks

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/AttributeDataFlow.cs:
```diff
@@ -0,0 +1,139 @@
+                TypeDesc td => new SystemTypeValue(td),
+                string str => new KnownStringValue(str),
+                // We shouldn't have gotten a None annotation from flow annotations since only string/Type can have annotations
+                _ => throw new InvalidOperationException()
```
This is reachable if the value is null. E.g. `[SomeAttribute(null)]`.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedCallGraph.cs:
```diff
@@ -0,0 +1,71 @@
+            toMembers.Add(toMember);
+        }
+
+        public void TrackCall(MethodDesc fromMethod, MethodDesc toMethod)
```
We should add an assert that fromMethod and toMethod are in typical forms. I assume this won't do the right thing if things are instantiated.

(Same comment for the methods below.)

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/EcmaExtensions.cs:
```diff
@@ -63,5 +65,51 @@ public static PropertyPseudoDesc GetProperty(this MetadataType mdType, string na
+                MetadataType type => type.ContainingType,
+                PropertyPseudoDesc property => property.OwningType,
+                EventPseudoDesc @event => @event.OwningType,
+                _ => null
```
I would make this assert/throw.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/FlowAnnotations.cs:
```diff
@@ -16,6 +16,8 @@
 
 using Debug = System.Diagnostics.Debug;
 using WellKnownType = Internal.TypeSystem.WellKnownType;
+using System.Reflection;
```
Nit: could you organize the usings the way they're organized (also in other files)?

MichalStrehovsky reviewed (commented):
I'm still looking through this, but this is how much I was able to look at today (I took a day off yesterday instead of today's public holiday).

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/FlowAnnotations.cs:
```diff
@@ -891,5 +923,43 @@ public FieldAnnotation(FieldDesc field, DynamicallyAccessedMemberTypes annotatio
+            }
+            else if (genericArgument is MetadataType genericArgumentType)
+            {
+                if (genericArgumentType.IsTypeOf(ILLink.Shared.TypeSystemProxy.WellKnownType.System_Nullable_T))
+                {
+                    var innerGenericArgument = genericArgumentType.Instantiation.Length == 1 ? genericArgumentType.Instantiation[0] : null;
+                    switch (innerGenericArgument)
+                    {
+                        case GenericParameterDesc gp:
+                            return new NullableValueWithDynamicallyAccessedMembers(genericArgumentType,
+                                new GenericParameterValue(gp, GetGenericParameterAnnotation(gp)));
+
+                        case TypeDesc underlyingType:
+                            return new NullableSystemTypeValue(genericArgumentType, new SystemTypeValue(underlyingType));
+                    }
+                }
+                // All values except for Nullable<T>, including Nullable<> (with no type arguments)
+                return new SystemTypeValue(genericArgumentType);
```
There are two main reasons for some of this weird "handlings", but it doesn't excuse most of them:
* Cecil's reference->definition resolution returns null if the resolution fails (missing assembly dependency, wrong dependency, ...). So unlike in NativeAOT's type system, there are potentially lot of places where null can appear unfortunately.
* Historically we've seen quite a few cases where weird IL/metadata we never anticipated (like obfuscated code) crashed the linker. While I agree it's easy to diagnose the problem, there's typically absolutely no workaround for the user (other than disabling trimming which in Blazor/Xamarin scenarios is not really viable).

We are guilty of adding `?.` and similar in places where we couldn't immediately reason about the input being null or not. This is exaggerated by the fact that we turned on nullable reference types (which was a good thing as it found several issues), but Cecil is not annotated, so everything looks nullable (even if in really it might not be).

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs:
```diff
@@ -0,0 +1,581 @@
+                        // Already warned above if multiple methods map to the same type
+                        // Fill in null for argument providers now, the real providers will be filled in later
+                        _generatedTypeToTypeArgumentInfo ??= new Dictionary<MetadataType, TypeArgumentInfo>();
+                        _ = _generatedTypeToTypeArgumentInfo.TryAdd(stateMachineType, new TypeArgumentInfo(method, null));
```
Nit: it's odd to call TryAdd and then discard the result.

```suggestion
                        _generatedTypeToTypeArgumentInfo[stateMachineType] = new TypeArgumentInfo(method, null);
```

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs:
```diff
@@ -0,0 +1,581 @@
+                    }
+                    var method = typeInfo.CreatingMethod;
+                    var body = ilProvider.GetMethodIL(method);
+                    if (body is not null)
```
Can we end up with a CreatingMethod that doesn't have a body?

vitek-karas replied:
I think this was added because of PInvokes - but I could be wrong - I'll try to remove it and see what happens.
I agree with you that it would be really weird.

vitek-karas replied:
I removed it and everything passed... so I guess it can't happen.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/EcmaExtensions.cs:
```diff
@@ -63,5 +65,51 @@ public static PropertyPseudoDesc GetProperty(this MetadataType mdType, string na
+                if (parameterMetadata.Index != index + 1)
+                    continue;
+
+                if (!method.Signature[index].IsByRef)
+                    return ReferenceKind.None;
```
Should this happen before we enter the foreach?

MichalStrehovsky reviewed (commented):
Looks great. Just a couple opportunities to delete some lines of code.

vitek-karas committed:
Port almost latest linker changes

This was needed to fix failures in libraries build with AOT. The underlying problem is not fully fixed: https://github.com/dotnet/linker/issues/2874
But with these changes it doesn't crash the compiler, it just leads to sometimes imprecise warnings.

vitek-karas commented:
The latest commit https://github.com/dotnet/runtime/pull/71485/commits/53800b7af8be348b767e1ea32b6dfa5761f72eae brings recent linker changes over. I suggest this is reviewed incrementally.

I didn't have a chance to apply PR feedback from Michal yet.

As is right now it passes smoke tests and libraries AOT build.

vitek-karas committed:
Merge branch 'main' into LinkerToAotSync2

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:
```diff
@@ -499,8 +497,12 @@ protected override void GetDependenciesDueToMethodCodePresenceInternal(ref Depen
-                    dependencies = dependencies ?? new DependencyList();
-                    dependencies.Add(factory.DataflowAnalyzedMethod(methodIL.GetMethodILDefinition()), "Method has annotated parameters");
+                    MethodIL methodILDefinition = methodIL.GetMethodILDefinition();
+                    if (!CompilerGeneratedState.IsNestedFunctionOrStateMachineMember(methodILDefinition.OwningMethod))
+                    {
```
@MichalStrehovsky I would welcome opinions on how to do this better. Basically we don't want to run data flow on compiler generated methods (as they are now processed as part of the parent method). Currently this change avoids adding the DataflowAnalyzedMethod node into the graph in this case, but it means that there are 3 places where we have to check for this.

Alternatives:
* Modify the DataflowAnalyzedMethod to "do nothing" in this case - cons, it would like as if we did data flow by just looking at the graph.
* Modify the factory.DataflowAnalyzedMethod - but this is tricky as it can't really return `null`
* Hide this inside `ReflectionMethodBodyScanner` - meaning the node will be added into the graph and it will trigger data flow, but the data flow will be a NOP.

MichalStrehovsky replied:
Add a `private void AddDataflowDependencyIfNeeded(ref DependencyList dependencies, NodeFactory factory, MethodIL methodIL, string reason)` method to `UsageBasedMetadataManager` and call it from the 3 places?

tlakollo commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMarker.cs:
```diff
@@ -135,40 +145,70 @@ void MarkEvent(in MessageOrigin origin, EventPseudoDesc @event, Origin memberWit
+                        MessageFormat.FormatRequiresAttributeUrlArg(DiagnosticUtilities.GetRequiresAttributeUrl(requiresAttribute.Value)));
+                }
+            }
+
```
The implementation in linker kind of calls CheckAndReportRequires here. Without it, very basic scenarios in reflection won't warn. For example, I was debugging today in the repro test the following:
``` C#
public static void Main ()
{
	typeof (BaseClassWithRequires).RequiresPublicMethods ();
}

class BaseClassWithRequires
{
	[RequiresUnreferencedCode ("Message")]
	[RequiresAssemblyFiles ("Message")]
	[RequiresDynamicCode ("Message")]
	public virtual void VirtualMethod ()
	{
	}
}
```
Or using the GetMethod() method which has the same effect of marking the method with Requires and not producing the warning.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/ILCompiler.Compiler.csproj:
```diff
@@ -20,6 +20,11 @@
     <ProjectReference Include="..\ILCompiler.MetadataTransform\ILCompiler.MetadataTransform.csproj" />
     <ProjectReference Include="..\ILCompiler.TypeSystem\ILCompiler.TypeSystem.csproj" />
+
+    <PackageReference Include="StaticCs" Version="$(StaticCsVersion)">
```
I think this will be a problem for source build.

It looks like this is just some analyzer with a custom attribute? I would be fine not running the analyzer in source build (or ever) and we can fake the custom attribute so that we don't have to make acrobatics in the shared files.

vitek-karas replied:
@agocke - could you please comment on this? I basically just copied this from linker, I don't know the details behind it.

agocke replied:
Yup, it's an analyzer and a source file. I think we can condition it out in source build.

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Logger.cs:
```diff
@@ -29,32 +32,40 @@ public class Logger
-        public Logger(ILogWriter writer, bool isVerbose, IEnumerable<int> suppressedWarnings, bool singleWarn, IEnumerable<string> singleWarnEnabledModules, IEnumerable<string> singleWarnDisabledModules)
+        public Logger(
+            ILogWriter writer,
+            CompilerGeneratedState compilerGeneratedState,
```
I had to keep the fact that `_compilerGeneratedState` can be null unfortunately - because the `Null` logger doesn't have `ILProvider` and the `CompilerGeneratedState` can't operate without one really.
Alternative would be to have an interface to the compiler generated state functionality needed by the logger and have a dummy implementation. Since the null only adds two places where it needs a null check it didn't feel that bad.

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs:
```diff
@@ -0,0 +1,581 @@
+                                case ILOpcode.call:
+                                case ILOpcode.callvirt:
+                                case ILOpcode.newobj:
+                                    lambdaOrLocalFunction = methodBody.GetObject(reader.ReadILToken()) as MethodDesc;
```
Good catch - thanks.

vitek-karas committed:
Mostly PR feedback and small cleanup

vitek-karas committed:
Merge branch 'LinkerToAotSync2' of https://github.com/vitek-karas/runtime into LinkerToAotSync2

vitek-karas committed:
Merge branch 'main' into LinkerToAotSync2

MichalStrehovsky commented:
For the `/_/src/libraries/System.Reflection/tests/AssemblyTests.cs(191,0): error IL3002: (NETCORE_ENGINEERING_TELEMETRY=Build) System.Reflection.Tests.AssemblyTests.<>c__DisplayClass20_0.<GetFile_InMemory>b__2(): Using member 'System.Reflection.Assembly.GetFiles()' which has 'RequiresAssemblyFilesAttribute' can break functionality when embedded in a single-file app. This member throws an exception for assemblies embedded in a single-file app.`, just add IL3002 to the list here:

https://github.com/dotnet/runtime/blob/81a3dcdad5235c38a329b372af132a0053dc1708/eng/testing/tests.singlefile.targets#L29

The test assemblies produce a lot of analysis warnings. We also don't care.

MichalStrehovsky reviewed (approved):
Thank you!

vitek-karas committed:
Ignore single-file warnings in AOT tests

vitek-karas committed:
Merge branch 'LinkerToAotSync2' of https://github.com/vitek-karas/runtime into LinkerToAotSync2

vitek-karas committed:
Fixes a bug when compiler generated code would not run DataFlow.

If the user method has no intersting code in it, but it has a lambda (for example) which does have an interesting code which needs data flow - we now need to add the user method for data flow analysis (since the user method and all of its compiler generated methods will be analyzed together).

This also adds a few smoke tests covering these scenarios.

vitek-karas commented:
The CoreCLR libraries tests are failing on https://github.com/dotnet/runtime/issues/72429 and  probably https://github.com/dotnet/runtime/issues/68289.
The extra platforms NativeAOT failure is https://github.com/dotnet/runtime/issues/72149.
